sycl/include/sycl/ext/intel/math.hpp

```diff
@@ -0,0 +1,69 @@
+//==------------- math.hpp - Intel specific math API -----------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// The main header of Intel specific math API
+//===----------------------------------------------------------------------===//
+
+#pragma once
+#include <sycl/half_type.hpp>
+#include <type_traits>
+
+// _iml_half_internal is internal representation for fp16 type used in intel
+// math device library. The definition here should align with definition in
+// https://github.com/intel/llvm/blob/sycl/libdevice/imf_half.hpp
+#if defined(__SPIR__)
+typedef _Float16 _iml_half_internal;
+#else
+typedef uint16_t _iml_half_internal;
+#endif
+
+extern "C" {
+float __imf_saturatef(float);
+float __imf_copysignf(float, float);
+double __imf_copysign(double, double);
+_iml_half_internal __imf_copysignf16(_iml_half_internal, _iml_half_internal);
+};
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext {
+namespace intel {
+namespace math {
+
+#if __cplusplus >= 201703L
+template <typename Tp> Tp saturate(Tp x) {
+  static_assert(std::is_same<Tp, float>::value,
+                "sycl::ext::intel::math::saturate only supports fp32 version.");
+  if constexpr (std::is_same<Tp, float>::value)
+    return __imf_saturatef(x);
+}
+
+template <typename Tp> Tp copysign(Tp x, Tp y) {
+  static_assert(std::is_same<Tp, float>::value ||
+                    std::is_same<Tp, double>::value ||
+                    std::is_same<Tp, sycl::half>::value,
+                "sycl::ext::intel::math::copysign only supports fp16, fp32, "
+                "fp64 version.");
+  if constexpr (std::is_same<Tp, float>::value)
+    return __imf_copysignf(x, y);
+  if constexpr (std::is_same<Tp, double>::value)
+    return __imf_copysign(x, y);
+  if constexpr (std::is_same<Tp, sycl::half>::value) {
+    static_assert(sizeof(sycl::half) == sizeof(_iml_half_internal),
+                  "sycl::half is not compatible with _iml_half_internal.");
+    _iml_half_internal xi = __builtin_bit_cast(_iml_half_internal, x);
+    _iml_half_internal yi = __builtin_bit_cast(_iml_half_internal, y);
+    return __builtin_bit_cast(sycl::half, __imf_copysignf16(xi, yi));
+  }
+}
```

Comment:
Could both of these be overloads instead, or is the intention to strictly disallow implicit conversions?

Reply:
Hi, @steffenlarsen 
I think your question is why don't use function overload for different version(fp16, fp32, fp64) of copysign?
I have tried function overloads like:
float copysign(float x) { return __imf_copysignf(x); }
double copysign(double x) { return __imf_copysign(x); }
sycl::half copysign(sycl::half) {...}

However, I found unnecessary undefined symbols would be involved into final host object files. For example, when user only calls "float copysign", all of __imf_copysignf/__imf_copysign/__imf_copysignf16 will be resided in host object file as "UND" symbols.
If we use template and compile-time if to remove unnecessary branch, we won't involve any unexpected symbols in host object file.

Thanks very much.

Reply:
That was indeed my question, and I see the problem. I am okay with the current version.